Add cyclic and reflected boundaries

cdcapano opened the pull request:
Kombine can have trouble with hard boundaries in large dimensional problems, especially in areas of the parameter space that are close to where multiple boundaries intersect (a hyperdimensional "corner"). The reason for this is kombine uses a KDE for its proposal density; since this smoothly varies to 0, the KDE can "leak" across boundaries. Depending on the shape of the parameter space and the priors, this may cause nearly all jump proposals to be outside of the space permitted by the prior. When that happens, all points are rejected, and a walker barely moves from its initial conditions. For example, note the outlier points close to the upper bound on the right ascension (alpha) in this run on GW150914:

https://www.atlas.aei.uni-hannover.de/~cdcapano/LSC/projects/pycbc_inference/cbc/gw150914/kombine/larger_dist-5Kwalkers_1Kiterations-500kde_pts/posteriors-scatter.png

I think (but not certain; have to test) that this may also be why Miriam's multi-mode ringdown runs using kombine return junk. None of the points are evolving, apparently because all of the jump proposals end up outside of the prior boundaries.

This patch tries to remedy this by introducing the ability to use cyclic and reflected boundaries, rather just the "hard" boundaries currently used. By cyclic, I mean that if a value exceeds the bounds, it will be wrapped back around. For example, if you have a uniform distribution between 0 and 2pi, and the jump proposal is for 3pi, then this will be mapped back to pi. By reflected, I mean that if value exceeds the bounds it will be reflected back by the amount it exceeds. So, if you have inclination between 0 and pi, and the jump proposal is 5pi/4, the value will be mapped to 3pi/4. I hope this might also be useful for parameters that have an unphysical boundary close to the peak of the log likelihood; say, if we sample in eta on a signal that is close to equal mass, putting a reflected boundary at eta = 0.25 might help.

As a first step, I've added a `boundaries_utils.py` module to inference that allows you to set different types of boundaries. Here are some plots of what cyclic and reflected boundaries at [-1,1) do to a set of input values between [-10, 10):
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/boundary_plots/cyclic_bounds.pdf
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/boundary_plots/reflect_left.pdf
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/boundary_plots/reflect_right.pdf
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/boundary_plots/reflected_well.pdf

I don't intend this patch to be merged yet; I just wanted to get the PR started. I'm currently adding the ability to distributions.py and to likelihood.py to make use of the `Bounds` class introduced here, which I'll update this PR with.

cdcapano commented:
Ok, this patch is just about ready to merge. I'm adding a hold label while I do a PE run to check that it works.

@stevereyes01 I addressed the issue you had with the Gaussian.rvs function not returning a structred array in this patch. In the process, I found that there was a bug in the Gaussian distribution; the pdf wasn't normalized correctly if bounds were provided. This patch fixes that.

Initializing a Gaussian is also a bit different. Instead of providing a tuple that includes the mean and the variance for each parameter, you just provide the bounds (if any). If you want to specify a mean and variance for the parameter, you add additional keyword arguments ending in `_mean` and `_var`. For example, the original way: `dist = Gaussian(mass1=(1,3,1.4,0.1))` is now `dist = Gaussian(mass1=(1,3), mass1_mean=1.4, mass1_var=0.1)`. If you don't provide a mean or var arg, it will default to zero mean with unit variance. Likewise, you can provide `None` for the bounds, in which case it will be an unbounded Gaussian.

Aside from that, this patch doesn't change the API for any of the other distributions. The difference is you can now specify what sort of bounds to use by adding `btype-(min|max)-{param} = reflected|circular|open|closed` to the ini file for pycbc inference. `UniformAngle` defaults to using circular bounds on `[0,2pi)` while `SinAngle` defaults to using reflected bounds on `[0,pi]`. By extension, `UniformSolidAngle` (`UniformSky`) default to being circular on `[0,2pi)` in the azimuthal angle (`ra`) and `[0,pi]` on the polar angle (`[-pi/2,pi/2]` in `dec`).

When the sampler passes parameter values to the likelihood evaluator, they are mapped to new values according to the boundary conditions before being passed to the waveform generator and prior functions. So, e.g., while the sampler will see `ra=2.1*pi`, the waveform generator and prior will see `ra=0.1*pi`. The mapped values are saved to the output file.

cdcapano commented:
Tested this doing a ringdown PE run. It doesn't cause errors, but it also doesn't fix the problem I was hoping it would fix, which is frustrating. Even so, since it adds some functionality and fixes some issues with the Gaussian distribution, I think it's ready to be merged.

stevereyes01 commented:
@cdcapano Thanks for addressing this. I'll make sure to note the changes you've made.

cdcapano commented:
@cmbiwer I made it so that reflected boundaries are not used as defaults for any of the priors, and added a warning to the user if `_BoundedDist` is initialized using reflected boundaries.

cmbiwer commented on pycbc/inference/distributions.py:
```diff
@@ -568,53 +699,41 @@ def rvs(self, size=1, param=None):
     Class Attributes
     ----------------
-    name : 'cos_angle'
+    name : 'sin_angle'
```
I think you accidently copied ``name : 'sin_angle'`` here. Can you change it back to ``name : 'cos_angle'``?

cmbiwer commented:
Ok, I had a look through this. Looks like ``_Bound`` feeds into ``Bounds`` which feeds into ``_BoundedDist`` which feeds into ``PriorEvaluator``. Looks fine except for what I think is a the typo in the docstring.

Just a comment that really has nothing to do with the implementation but is there any reason to call it ``boundary_utils.py`` instead just ``bounds.py`` or ``boundary.py``. For the most part in PyCBC we aren't appending suffices to the ends of modules. Just a style thing but it sort of sticks out.

If you address the typo in the docstring, this is fine to merge.

cmbiwer commented:
Also as far as the QuantifiedCode issues, the first one about the wildcard I don't really think is an issue. We do this all over ``__init__.py`` in PyCBC. And the other ones look like they're all ``NotImplementedError`` placeholder functions. So I don't think these are an issue.

ahnitz commented:
For what its worth, I agree on the naming of modules with Chris '_utils'
seems superfluous.

On Fri, Nov 18, 2016 at 10:33 PM, Christopher M. Biwer <
notifications@github.com> wrote:

> Also as far as the QuantifiedCode issues, the first one about the wildcard
> I don't really think is an issue. We do this all over __init__.py in
> PyCBC. And the other ones look like they're all NotImplementedError
> placeholder functions. So I don't think these are an issue.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ligo-cbc/pycbc/pull/1149#issuecomment-261647553>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACGrRjkbDEChRThQbgmPUOPK6EW_tyOOks5q_hmggaJpZM4KWZG1>
> .
>

cdcapano commented:
@cmbiwer @ahnitz Ok, I fixed the doc string, and renamed `boundary_utils.py` to `boundaries.py` (I thought about `bounds.py`, but there are several places in `distributions.py` where variables called `bounds` are initialized).

cdcapano commented:
@cmbiwer Are you happy to merge this?

cmbiwer commented:
Ah sorry, I thought you were going to merge after those commits. Yes, I'm fine with it. I'll merge it.
